numpy/core/src/common/simd/avx512/arithmetic.h

```diff
@@ -112,5 +115,64 @@ NPY_FINLINE __m512i npyv_mul_u8(__m512i a, __m512i b)
 // TODO: emulate integer division
 #define npyv_div_f32 _mm512_div_ps
 #define npyv_div_f64 _mm512_div_pd
+NPY_FINLINE float npyv__sum_f32(__m256 a)
+{
+    __m128 t1 = _mm_add_ps(_mm256_castps256_ps128(a), _mm256_extractf128_ps(a,1));
+    __m128 t2 = _mm_movehdup_ps(t1);
+    __m128 t3 = _mm_add_ps(t1, t2);
+    __m128 t4 = _mm_movehl_ps(t3, t3);
+    __m128 t5 = _mm_add_ss(t3, t4);
+    return _mm_cvtss_f32(t5);
+}
+
+NPY_FINLINE double npyv__sum_f64(__m256d a)
+{
+    __m128d t1 = _mm_add_pd(_mm256_castpd256_pd128(a), _mm256_extractf128_pd(a,1));
+    __m128d t2 = _mm_unpackhi_pd(t1, t1);
+    __m128d t3 = _mm_add_sd(t2, t1);
+    return _mm_cvtsd_f64(t3);
+}
+
+NPY_FINLINE __m256 get_low_f32(__m512 a)
+{
+    return _mm512_castps512_ps256(a);
+}
+
+NPY_FINLINE __m256 get_high_f32(__m512 a)
+{
+#ifdef NPY_HAVE_AVX512DQ
+    return _mm512_extractf32x8_ps(a, 1);
+#endif
+    return _mm256_castpd_ps(_mm512_extractf64x4_pd(_mm512_castps_pd(a),1));
+}
+
+NPY_FINLINE __m256d get_low_f64(__m512d a)
+{
+    return _mm512_castpd512_pd256(a);
+}
+
+NPY_FINLINE __m256d get_high_f64(__m512d a)
+{
+    return _mm512_extractf64x4_pd(a, 1);
+}
+
+// Horizontal add: Calculates the sum of all vector elements.
+NPY_FINLINE float npyv_sum_f32(__m512 a)
+{
+#if defined(__INTEL_COMPILER)
+    return _mm512_reduce_add_ps(a);
+#else
+    return npyv__sum_f32(_mm256_add_ps(get_low_f32(a), get_high_f32(a)));
+#endif
+}
+
+NPY_FINLINE double npyv_sum_f64(__m512d a)
+{
+#if defined(__INTEL_COMPILER)
+    return _mm512_reduce_add_pd(a);
+#else
+    return npyv__sum_f64(_mm256_add_pd(get_low_f64(a), get_high_f64(a)));
+#endif
+}
```

Comment:
```suggestion
NPY_FINLINE float npyv_sum_f32(npyv_f32 a)
{
    __m512 h64   = _mm512_shuffle_f32x4(a, a, _MM_SHUFFLE(3, 2, 3, 2));
    __m512 sum32 = _mm512_add_ps(a, h64);
    __m512 h32   = _mm512_shuffle_f32x4(sum32, sum32, _MM_SHUFFLE(1, 0, 3, 2));
    __m512 sum16 = _mm512_add_ps(sum32, h32);
    __m512 h16   = _mm512_permute_ps(sum16, _MM_SHUFFLE(1, 0, 3, 2));
    __m512 sum8  = _mm512_add_ps(sum16, h16);
    __m512 h4    = _mm512_permute_ps(sum8, _MM_SHUFFLE(2, 3, 0, 1));
    __m512 sum4  = _mm512_add_ps(sum8, h4);
    return _mm512_cvtss_f32(sum4);
}
NPY_FINLINE double npyv_sum_f64(npyv_f64 a)
{
    __m512d h64   = _mm512_shuffle_f64x2(a, a, _MM_SHUFFLE(3, 2, 3, 2));
    __m512d sum32 = _mm512_add_pd(a, h64);
    __m512d h32   = _mm512_permutex_pd(sum32, _MM_SHUFFLE(1, 0, 3, 2));
    __m512d sum16 = _mm512_add_pd(sum32, h32);
    __m512d h16   = _mm512_permute_pd(sum16, _MM_SHUFFLE(2, 3, 0, 1));
    __m512d sum8  = _mm512_add_pd(sum16, h16);
    return _mm512_cvtsd_f64(sum8);
}
```

Reply:
This implementation almost the same as `_mm512_reduce_add_ps/pd` but works all for all compilers.
I avoided the other technique that based on horizontally-add since `haddps/vhaddps/haddps/vhaddpd` became more expensive(latency) to use started from Skylake.